### ISSUE TITLE
.staging turds everywhere

### ISSUE THREAD

**Issue by cbarraford** [nanovms/ops#135]
I noticed that whenever i run the `ops` command, a `.staging` directory is created in the pwd, which leaves `.staging` directories all over the place on my filesystem. Its surprising that running a command like `ops list` creates directories and in the pwd.

Is this on purpose? I can imagine linux purist getting pissy about the uncleanliness of this.

**Comment by anikitenko** [comment]
I'm currently working on implementing solution to not use ".staging" in each directory. The only thing I'm wondering is what means error `assertion n failed in /home/tijoytom/nanovms/nanos/src/runtime/metadata.h: resolve_path() on line 44; halt`? This is happening during `load` but `run` works properly. BTW, I checked all funcs and checked what may cause problem, checked manifest, checked `TargetRoot`, ran with verbose and debug modes.. didn't find anything.. :(
So my question is what's happening in `resolve_path` function that causes an error?
Also JFI: tested with php_7.2.13

**Comment by tijoytom** [comment]
What the command line that you used?

**Comment by anikitenko** [comment]
Also forgot to mention, why I decided to implement this.. After implementing solution we'll be able to control what we have locally and one of interesting things is to be able to implement #147

**Comment by anikitenko** [comment]
So.. I used `ops load php_7.2.13 -a ex.php -d -v` and the following output I saw:
`Extracting /Users/alexvwan/.ops/packages/php_7.2.13.tar.gz...
[php ex.php]
booting image ...
qemu-system-x86_64 -drive file=/var/folders/vg/59fslhy12yg81x72t44twzt80000gn/T/.ops/staging_112407492/image,format=raw,if=virtio -device virtio-net,netdev=n0 -netdev user,id=n0 -nodefaults -no-reboot -device isa-debug-exit -m 2G -display none -serial stdio
assertion n failed in /home/tijoytom/nanovms/nanos/src/runtime/metadata.h: resolve_path() on line 44; halt
exit status 1
`

**Comment by tijoytom** [comment]
You probably still want the .staging for the final image, but can move boot.img, stage3.img and mkfs to ./ops folder.

**Comment by eyberg** [comment]
i'm in favor of nuking these .staging folders - there's no reason to keep them around

@anikitenko - there was a recent change in mkfs - we are in the process of putting in some versioning for ops && such but for now -- maybe try nuking ~/.ops && re-downloading to see if this gets you the newer bins

**Comment by anikitenko** [comment]
Regarding moving boot.img, stage3.img and mkfs, I have PR #195 and I'm in progress..
In solution that I'm implementing final image, which is `image`, `tempimage` and extracted `php_7.2.13` are in one place in temp directory. During run it's using correct path and arguments to `qemu` are correct. But why function `resolve_path()` returns error.. I don't know..

**Comment by anikitenko** [comment]
@eyberg I already tried (and just now) to re download mkfs, the error still exists.. :(

**Comment by eyberg** [comment]
@anikitenko - can you help - at - nanovms.com with the following information:

* copy of that ex.php if you can

output of this ->
```
 md5 .staging/mkfs
  md5 .staging/stage3.img
md5 .staging/boot.img
```

i have a feeling one of these is not matching up but we'll see

**Comment by anikitenko** [comment]
@eyberg, sure:
* copy of that ex.php:
`<?php
      phpinfo();
      ?>
`
* md5 .staging/mkfs:
`MD5 (.staging/mkfs) = 83cabf887cbcfd3dffa95ac920876ac5`
* md5 .staging/stage3.img:
`MD5 (.staging/stage3.img) = 0fa3c5038c3bf6aae4497952e4a156f4`
* md5 .staging/boot.img:
`MD5 (.staging/boot.img) = fab7aa4a884c6705b9274f51c29c1873`

**Comment by eyberg** [comment]
@anikitenko - you're using 'ops run' w/a local php or is that for a diff non-php project?

**Comment by anikitenko** [comment]
@eyberg when I run code with my changes using command `run` it works as expected. Problem is happening when I'm trying to use `load`. The command is `~/dev/src/github.com/nanovms/ops/ops load php_7.2.13 -a ex.php -d -v`. And I'm running without local php.

**Comment by eyberg** [comment]
well the ```ops run``` command would be pointed at a binary right? what binary is that? is that a php interpreter or something else? - the reason I'm asking is that this is a mkfs error complaining about something on the fs image it's generating and maybe mixing these two together is populating files it might not expect (not sure just guessing)

**Comment by anikitenko** [comment]
Right, I compiled binary as in example https://nanovms.gitbook.io/ops/examples
Hm.. I think, will it be better if I open pull request with my changes, so you can see what has been changed and it will help to figure out what caused that error?

**Comment by eyberg** [comment]
yeh - that'd be great - pls do so we can keep this tkt focus'd on '.staging turds' ;)

**Comment by anikitenko** [comment]
@eyberg, I created PR with whole description about "what was an idea of this PR" #208

**Comment by anikitenko** [comment]
Finally I was able to resolve the problem.. It appears that func `addFilesFromPackage` was working assuming that package will be extracted to ".staging" directory. Function walks through relative path to package, for example `.staging/php_7.2.13` and was splitting it. If 3rd argument is `sysroot` then it should go to root(/) ([.staging, php_7.2.13, sysroot]) everything else to /package_name(php_7.2.13)/*.
So.. since I changed path to temp one, split was working incorrectly and files were not added to image.

I changed that func `addFilesFromPackage` and now it working :) Now it does the following:
* First walk through const PackageRootPath which is `sysroot` in package and add all files to root(/) in image
* Second walk through everything else, except PackageRootPath and add to /package_name(php_7.2.13)/*. package_name is determined by getting `filepath.Base` from path to extracted package

And I updated PR..
